### PR TITLE
Add `filetype` to Science Museum script

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Related to #536 by @obulat 

## Description

This PR adds `filetype` and `filesize` to the list of data collected from the Science Museum API.

## Testing Instructions

The tests for the script should pass.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 1]
```diff
@@ -171,19 +173,28 @@ def _get_creator_info(obj_attr):
...
 
 
+def _get_filesize(image_url):
+    """
+    Get the size of the image in bytes.
+    """
+    resp = delayed_requester.get(image_url)
+    if resp:
+        filesize = int(resp.headers.get("Content-Length", 0))
+        return filesize if filesize != 0 else None
```
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:183]: If I'm understanding this correctly, this will make a request for every image we receive from their API. Should this maybe go the way of NYLP and not report `filesize` since it's not provided by their API? And if not, should this maybe be doing a `HEAD` request rather than a full `GET`?

File: tests/dags/providers/provider_api_scripts/test_science_museum.py [hunk 2]
```diff
@@ -156,41 +156,69 @@ def test_creator_info_fail():
...
+        actual_filesize,
+    ) = sm._get_image_info(large_image)
     expected_image = (
         "https://coimages.sciencemuseumgroup.org.uk/images/3/563/"
         "large_1999_0299_0001__0002_.jpg"
     )
     expected_height = 1022
     expected_width = 1536
+    expected_filetype = "jpeg"
+    expected_filesize = None
```
Comment (AetherUnbound) [inline tests/dags/providers/provider_api_scripts/test_science_museum.py:173]: Would it be possible to add a test case where filesize is a non-`None` value?

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py (hunk 1 above)
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:183]: I agree with not reporting `filesize` at all, this is a more sustainable approach.
By the way, I was sure I was using the `HEAD` requests everywhere, and not `GET`'s 😵‍💫

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 3]
```diff
@@ -173,60 +172,51 @@ def _get_creator_info(obj_attr):
...
-    if measurements:
-        dimensions = measurements.get("dimensions")
-        if dimensions:
-            for dim in dimensions:
-                height_width[dim.get("dimension")] = dim.get("value")
-    return height_width.get("height"), height_width.get("width")
+def _get_dimensions(image_data: Dict) -> Tuple[int | None, int | None]:
+    """
+    Returns the height and width of the image from a list of dictionaries
+    with keys of "dimension", "units", "value".
```
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:202]: This is not super important here, but we could use a [`TypedDict`](https://peps.python.org/pep-0589/) for cases like this! Also, it seems like it has a `measurements` key with a `dimensions` key inside the dictionary there?

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 4]
```diff
@@ -173,60 +172,51 @@ def _get_creator_info(obj_attr):
...
+def _get_license_version(source: Dict | None) -> str | None:
     license_version = None
     if source:
-        legal = source.get("legal")
-        if legal:
-            rights = legal.get("rights")
-            if type(rights) == list:
-                license_version = rights[0].get("usage_terms")
+        rights = source.get("legal", {}).get("rights")
+        if type(rights) == list and len(rights) > 0:
```
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:218]: We should use `isinstance` here because it's [faster and will respect inheritance](https://switowski.com/blog/type-vs-isinstance)!

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 5]
```diff
@@ -171,51 +172,51 @@ def _get_creator_info(obj_attr):
...
-        if dimensions:
-            for dim in dimensions:
-                height_width[dim.get("dimension")] = dim.get("value")
-    return height_width.get("height"), height_width.get("width")
+def check_url(image_url: str | None) -> str | None:
+    if not image_url:
+        return None
+    if image_url.startswith("http"):
+        return image_url
+    else:
```
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:195]: This `else` could probably be dropped in this case!

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py (hunk 3 above)
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:202]: Thank you for the link, it looks interesting! Can we leave it for the next refactor 😉

I've changed the signature of this function. Instead of trying to get the `measurements` in one of two places using `measurements = processed.get("large / medium").get("measurements")`, I call the `_get_dimensions` with `image_data` we get using ``measurements = processed.get("large / medium")`, and then extract the measurements inside this function:
`height, width = _get_dimensions(image_data)`

Comment (AetherUnbound) [reply openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:202]: Yes, another time on that one! Just wanted to mention it :) 

And that works great! But I think the documentation should still read `with keys of "measurement"."dimensions", "units", and "value".` (since the top level key is still `measurements` rather than `dimensions`).

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 6]
```diff
@@ -173,60 +173,50 @@ def _get_creator_info(obj_attr):
...
+def _get_license_version(source: Dict | None) -> str | None:
     license_version = None
     if source:
-        legal = source.get("legal")
-        if legal:
-            rights = legal.get("rights")
-            if type(rights) == list:
-                license_version = rights[0].get("usage_terms")
+        rights = source.get("legal", {}).get("rights")
+        if isinstance(rights, list) and len(rights) > 0:
```
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:218]: Nice simplification! Maybe this could even be?
```suggestion
        rights = source.get("legal", {}).get("rights", [])
        if rights:
```

This defaults to a list and just does a truthy check on the list to see if it has items. I guess the `isinstance` check would be important if `rights` could ever be an truthy object that's _not_ a list, like a dict or something 🤔

**Review by AetherUnbound — APPROVED** [review]
Just a couple more notes, otherwise this looks good!

**Comment by obulat** [comment]
> And that works great! But I think the documentation should still read with keys of "measurement"."dimensions", "units", and "value". (since the top level key is still measurements rather than dimensions).

Ah, I didn't even see the docstring! 🫢

File: openverse_catalog/dags/providers/provider_api_scripts/science_museum.py [hunk 7]
```diff
@@ -171,51 +173,50 @@ def _get_creator_info(obj_attr):
...
+    if not image_url:
+        return None
+    if image_url.startswith("http"):
+        return image_url
+    return f"https://coimages.sciencemuseumgroup.org.uk/images/{image_url}"
+
+
+def _get_dimensions(image_data: Dict) -> Tuple[int | None, int | None]:
+    """
+    Returns the height and width of the image from a list of dictionaries
```
Comment (obulat) [inline openverse_catalog/dags/providers/provider_api_scripts/science_museum.py:201]:
```suggestion
    Returns the height and width of the image from "image_data"."measurements"
```

**Comment by obulat** [comment]
I am going to draft this for now so that the reviewers are not bothered, and if the new Provider Data Ingester is merged today, I'll move the work from here to #576.
